sstables/sstable_set.cc

```diff
@@ -331,6 +343,16 @@ stop_iteration partitioned_sstable_set::for_each_sstable_until(std::function<sto
     return stop_iteration::no;
 }
 
+future<stop_iteration> partitioned_sstable_set::for_each_sstable_gently_until(std::function<future<stop_iteration>(const shared_sstable&)> func) const {
+    for (auto& sst : *_all) {
+        auto stop = co_await func(sst);
```

Comment:
If the sstable_set is modified during a yield, the iteration will break.

Reply:
This is currently called with compaction disabled (temporarily) so nothing will be deleted from the set
but it is a good point nevertheless.

We can copy the vector aside and traverse the vector.
Although this won't reflect any concurrent deletion (or additions) to the set, but it's not a problem at this level.

Another option is to use a rwlock, which I'm not thrilled about.

@raphaelsc what do you think?

Reply:
Also, the current for_each_sstable function seems synchronous, but in fact it is being called in a seastar thread on the functors passed to it call maybe_yield too, so the problem is preexisting.

See for example https://github.com/scylladb/scylladb/blob/550aa012429f8832105c26f0f672626ebd78c968/compaction/compaction_manager.cc#L1620-L1625

Given this works since it is called under disabled compaction I think that a rwlock will provide the required protection while adding only very little overhead.

Reply:
scratch the rwlock idea. insert is synchronous and may invalidate the for_each iterators if e.g. a vector is reallocated.

I'll send a separate fix to this preexisting issue (confined to master)

Reply:
we workaround concurrent updates by not modifying a sstable set in place. that's important for concurrent reads, and cache which references the sstable set as snapshot source.

I don't think this needs to be enforced by the sstable_set interface. 

The same way that iterating through a std::list can will race with concurrent updates. The responsibility is pushed to the user.

Reply:
You're right about that.
But that said, we still need must hold on to the `lw_shared_ptr<sstables::sstable_set>`
if we plan to yield while in for_each_sstable, otherwise it may go away when
it is replaced, e.g. in `compaction_group::set_main_sstables`.

Reply:
https://github.com/scylladb/scylladb/pull/13852

Reply:
v2 is rebased on top of the above so the loop calling update_sstable_cleanup_state gently keeps the set alive using a shared pointer even if the loop preempts.